espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/adapter/DummyW3CActionAdapter.kt

```diff
@@ -0,0 +1,132 @@
+import io.appium.espressoserver.lib.helpers.w3c.state.KeyInputState
+import io.appium.espressoserver.lib.helpers.w3c.state.PointerInputState
+
+open class DummyW3CActionAdapter : BaseW3CActionAdapter() {
+
+    // Keep a log of pointer move events so the values can be checked in the unit tests
+    private val pointerMoveEvents = ArrayList<PointerMoveEvent>()
+    override val logger: Logger = DummyLogger()
+
+    override val viewportHeight: Long
+        get() = 400
+
+    override val viewportWidth: Long
+        get() = 200
+
+    class PointerMoveEvent {
+        var sourceId: String? = null
+        var pointerType: PointerType? = null
+        var currentX: Float = 0.toFloat()
+        var currentY: Float = 0.toFloat()
+        var x: Float = 0.toFloat()
+        var y: Float = 0.toFloat()
+        var buttons: Set<Int>? = null
+        var globalKeyInputState: KeyInputState? = null
+    }
+
+    private inner class DummyLogger : Logger {
+        override fun error(vararg messages: Any) {
+            // No-op
+        }
+
+        override fun error(message: String, throwable: Throwable) {
+            // No-op
+        }
+
+        override fun info(vararg messages: Any) {
+            // No-op
+        }
+
+        override fun debug(vararg messages: Any) {
+            // No-op
+        }
+
+        override fun warn(vararg messages: Any) {
+            // No-op
+        }
+    }
+
+    override fun keyDown(keyEvent: W3CKeyEvent) {
+        // No-op
+    }
+
+    override fun keyUp(keyEvent: W3CKeyEvent) {
+        // No-op
+    }
+
+    @Throws(AppiumException::class)
+    override fun pointerUp(button: Int?, sourceId: String?, pointerType: PointerType?,
+                       x: Float?, y: Float?, depressedButtons: Set<Int>,
+                       globalKeyInputState: KeyInputState?) {
+        // No-op
+    }
+
+    @Throws(AppiumException::class)
+    override fun pointerDown(button: Int?, sourceId: String?, pointerType: PointerType?,
+                         x: Float?, y: Float?, depressedButtons: Set<Int>,
+                         globalKeyInputState: KeyInputState?) {
+        // No-op
+    }
+
+    override fun pointerCancel(sourceId: String, pointerType: PointerType) {
+        // No-op
+    }
+
+    override fun getPointerMoveDurationMargin(pointerInputState: PointerInputState): Double {
+        return if (pointerInputState.type == PointerType.TOUCH && !pointerInputState.hasPressedButtons()) {
+            // If no buttons are pushed nothing happens, so skip to the end
+            // of the pointer move
+            // e.g.: touch move without pressed buttons is like a finger moving without
+            //      being pressed on the screen
+            1.0
+        } else 0.01
+
+        // Give a margin of error of 1%
+    }
+
+    override fun pointerMove(sourceId: String?, pointerType: PointerType?,
+                             currentX: Float?, currentY: Float?, x: Float?, y: Float?,
```

Comment:
should these values be optional?

Reply:
Yeah, I changed it all to non-optional (except PointerInput which can be null).